### PR TITLE
fix: allow group names with commas

### LINKED ISSUE #3138: Cannot use comma in group names
## Describe the bug
If a group synchronized from an identity provider has a comma in the name it causes an error on login.

## Expected behavior
Login succeeds and the group is added with an escaped comma.

### PR DESCRIPTION
**Pull request by BruceMacD**
## Summary
Update group tracking to use a JSON binary in the database, rather than a text field separated by commas. This allows for storing group commas.

## Checklist

- [x] Wrote appropriate unit tests
- [x] Considered security implications of the change
- [x] Updated associated docs where necessary
- [x] Change is backwards compatible if it needs to be (user can upgrade without manual steps?)
- [x] Nothing sensitive logged
- [x] Considered data migrations for smooth upgrades

## Related Issues

Resolves #3138

### CODE REVIEW

File: internal/server/data/migrations.go [hunk 1]
```diff
@@ -1329,3 +1330,52 @@ func addAccessKeyIssuedForKind() *migrator.Migration {
...
+		Migrate: func(tx migrator.DB) error {
+			var groupType string
+			if err := tx.QueryRow(`SELECT pg_typeof(groups) FROM provider_users LIMIT 1;`).Scan(&groupType); err != nil {
+				return err
+			}
+			if groupType == "jsonb" {
+				// this migration has already been performed
+				return nil
+			}
+			if _, err := tx.Exec(`ALTER TABLE provider_users ADD COLUMN IF NOT EXISTS groups_json jsonb;`); err != nil {
```
Comment (mxyng) [inline internal/server/data/migrations.go:1346]: Does this need a temporary column? It seems you can retrieve the user groups, change the column, then reinsert the rows, e.g. add ALTER TABLE between the scanRows on line 1359 and the for loop on line 1366

Comment (BruceMacD) [reply internal/server/data/migrations.go]: Good observation, made this change.

File: internal/server/data/migrations.go [hunk 2]
```diff
@@ -1329,3 +1330,48 @@ func addAccessKeyIssuedForKind() *migrator.Migration {
...
+			users, err := scanRows(rows, func(u *userGroups) []any {
+				return []any{&u.IdentityID, &u.ProviderID, &u.Groups}
+			})
+			if err != nil {
+				return err
+			}
+			if _, err := tx.Exec(`ALTER TABLE provider_users DROP COLUMN IF EXISTS groups;`); err != nil {
+				return err
+			}
+			if _, err := tx.Exec(`ALTER TABLE provider_users ADD COLUMN IF NOT EXISTS groups jsonb;`); err != nil {
```
Comment (mxyng) [inline internal/server/data/migrations.go:1365]: You should be able to use a single `ALTER TABLE`

Comment (BruceMacD) [reply internal/server/data/migrations.go:1365]: Just tried this and it turns out you can't drop and add in a single statement.

**Review by mxyng — APPROVED** [review]

File: internal/server/data/schema.sql [hunk 3]
```diff
@@ -259,15 +259,15 @@ CREATE TABLE provider_users (
...
     last_update timestamp with time zone,
     redirect_url text,
     access_token text,
     refresh_token text,
     expires_at timestamp with time zone,
     given_name text DEFAULT ''::text,
     family_name text DEFAULT ''::text,
-    active boolean DEFAULT true
+    active boolean DEFAULT true,
+    groups jsonb
```
Comment (jmorganca) [inline internal/server/data/schema.sql:270]: nice choice of `jsonb`

**Review by jmorganca — APPROVED** [review]

**Pull request merged**